Tests/MathLib/TestGaussLegendreIntegration.cpp

```diff
@@ -0,0 +1,585 @@
+    double getAnalyticalIntegralOverUnitCube() const override
+    {
+        double const a = -.5;
+        double const b = .5;
+
+        double res = 1.0;
+        for (unsigned d : {0, 1, 2})
+        {
+            double poly = 0.0;
+            for (unsigned n = 0; n <= _degree; ++n)
+            {
+                poly += coeffs[n + d * (_degree + 1)] *
+                        (std::pow(b, n + 1) - std::pow(a, n + 1)) / (n + 1);
+            }
+
+            res *= poly;
+        }
+
+        return res;
+    }
+
+private:
+    unsigned const _degree;
+};
+
+unsigned binomial_coefficient(unsigned n, unsigned k)
+{
+    EXPECT_GE(n, k);
+    unsigned res = 1;
+
+    for (unsigned i = n; i > k; --i)
+    {
+        res *= i;
+    }
+
+    for (unsigned i = n - k; i > 0; --i)
+    {
+        res /= i;
+    }
+
+    return res;
+}
+
+/* This function is a polynomial where for each monomial a_ijk x^i y^j z^k
+ * holds: i + j + k <= n, where n is the overall polynomial degree
+ */
+struct F3DNonseparablePolynomial final : FBase
+{
+    // The number of coefficients/monomials are obtained as follows: Compute the
+    // number of combinations with repititions when drawing
+    // polynomial_degree times from the set { x, y, z, 1 }
+    F3DNonseparablePolynomial(unsigned polynomial_degree)
+        : FBase(binomial_coefficient(4 + polynomial_degree - 1, 4 - 1)),
+          _degree(polynomial_degree)
+    {
+    }
+
+    double operator()(std::array<double, 3> const& coords) const override
+    {
+        auto const x = coords[0];
+        auto const y = coords[1];
+        auto const z = coords[2];
+
+        double res = 0.0;
+        unsigned index = 0;
+
+        for (unsigned x_deg = 0; x_deg <= _degree; ++x_deg)
+        {
+            for (unsigned y_deg = 0; x_deg + y_deg <= _degree; ++y_deg)
+            {
+                for (unsigned z_deg = 0; x_deg + y_deg + z_deg <= _degree;
+                     ++z_deg)
+                {
+                    EXPECT_GT(coeffs.size(), index);
+
+                    res += coeffs[index] * std::pow(x, x_deg) *
+                           std::pow(y, y_deg) * std::pow(z, z_deg);
+
+                    ++index;
+                }
+            }
+        }
+
+        EXPECT_EQ(coeffs.size(), index);
+
+        return res;
+    }
+
+    double getAnalyticalIntegralOverUnitCube() const override
+    {
+        double const a = -.5;
+        double const b = .5;
+
+        double res = 0.0;
+        unsigned index = 0;
+
+        for (unsigned x_deg = 0; x_deg <= _degree; ++x_deg)
+        {
+            for (unsigned y_deg = 0; x_deg + y_deg <= _degree; ++y_deg)
+            {
+                for (unsigned z_deg = 0; x_deg + y_deg + z_deg <= _degree;
+                     ++z_deg)
+                {
+                    EXPECT_GT(coeffs.size(), index);
+
+                    res += coeffs[index] *
+                           (std::pow(b, x_deg + 1) - std::pow(a, x_deg + 1)) /
```

Comment:
minor: better to use `boost:: ...::pow<x_deg + 1>` for sake of speed, since the exponents are integral.